### ISSUE TITLE
Images with display:inline styles fail to render

### ISSUE THREAD

**Issue by westonruter** [ampproject/amp-wp#1803]
Originally reported at https://wordpress.org/support/topic/some-images-not-loading-version-1-0/page/2/

Given post content that looks like this:

```html
<img src="https://via.placeholder.com/400" width=400 height=400 style="display:inline">
```

Which is converted into AMP as:

```html
<style amp-custom>.amp-wp-5677b98{display:inline}</style>
...
<p><amp-img src="https://via.placeholder.com/400" width="400" height="400" class="amp-wp-enforced-sizes amp-wp-5677b98" layout="intrinsic"></amp-img></p>
```

The image will fail to render in AMP. According to @sebastianbenz this is because the `display:inline` conflicts with the AMP layout.

This issue did not happen in 0.7 because prior to 1.0 the `style` attribute was not copied from the `img` to the `amp-img`. This changed as of https://github.com/ampproject/amp-wp/pull/1121/commits/f9f3a054d7d9c66c935fbf4271cd4bc91679479c.

I'm not entirely sure what the best way to fix this is. Naturally we need the styles to persist from non-AMP to AMP for design fidelity. But the issue is not just present when there is an inline `style` attribute. The issue is also present if in the non-AMP version of the page there is markup like:

```html
<style>
img#inlined { display:inline !important; }
</style>
<img id="inlined" src="https://via.placeholder.com/400" width=400 height=400>
```

The only fix I can think of at the moment is to add CSS like the following to every page:

```css
:root:not(#_):not(#_):not(#_):not(#_):not(#_):not(#_):not(#_):not(#_):not(#_):not(#_) amp-img {
	display: inline-block;
}
```

But this would also be bad because we'd only want to select images that have `display:inline`; this would mess up other valid `display` values.

**Comment by randomascii** [comment]
Does it make sense to start by just fixing this for the style:inline case? That style can safely be deleted because it is the default, and it appears to be automatically inserted by some blogging software such as Open Live Writer (and perhaps Windows Live Writer as well)

**Comment by westonruter** [comment]
@randomascii Yes, perhaps. To be safe should probably replace `display:inline` with `display:inline-block`, just in case there are style rules applying a different `display`.

Do you have some examples of the raw HTML that Open Live Writer produces?

**Comment by randomascii** [comment]
I just restored the known-problematic post from last week. The URL is:

https://randomascii.wordpress.com/2019/04/21/on2-in-createprocess/

The problematic sequence, which appears four times, is shown here:

<p><a href="https://randomascii.files.wordpress.com/2019/04/image-2.png"><img width="631" height="452" title="image" style="display:inline;background-image:none;" alt="image" src="https://randomascii.files.wordpress.com/2019/04/image_thumb-2.png?w=631&#038;h=452" border="0"></a></p>

**Comment by randomascii** [comment]
Huh - that's interesting. Put another way, the contents of a typical img tag is:

img width="631" height="452" title="image" style="display:inline;background-image:none;" alt="image" src="https://randomascii.files.wordpress.com/2019/04/image_thumb-2.png?w=631&#038;h=452" border="0"

**Comment by westonruter** [comment]
I formatted your comment with backtick code blocks to prevent the HTML from being rendered.

So in this case, if the `style` attribute uses `display:inline-block`, there would be no negative side-effects from your perspective, right? I tried replacing it in DevTools on your page, and it seemed fine.

**Comment by randomascii** [comment]
That sounds good. I think that removing display:inline would also work, but I'm not really qualified to say. Thanks for looking.

**Comment by westonruter** [comment]
Note that this is not just a problem with images. It seems to be an issue with any element with a `layout` other than `fill` (and `nodisplay` naturally). 

Also, replacing `display:inline` with `display:inline-block` can't be done universally. In particular, if you have an element with `layout=fixed-height` giving it `display:inline-block` will actually break it, as this needs to be `display:block`.

Basically, we need to prevent rules such as these from being overridden:

```css
.i-amphtml-layout-container, .i-amphtml-layout-fixed-height, [layout=container], [layout=fixed-height][height] { display: block; }
.i-amphtml-layout-fixed, [layout=fixed][width][height]:not(.i-amphtml-layout-fixed) { display: inline-block; }
```

**Comment by westonruter** [comment]
Here's a more detailed list of the AMP style rules that apply to layout:

```css
.i-amphtml-element {
    display: inline-block
}

[layout=nodisplay]:not(.i-amphtml-element) {
    display: none !important
}

.i-amphtml-layout-fixed,
[layout=fixed][width][height]:not(.i-amphtml-layout-fixed) {
    display: inline-block;
    position: relative
}

.i-amphtml-layout-responsive,
[layout=responsive][width][height]:not(.i-amphtml-layout-responsive),
[width][height][sizes]:not(.i-amphtml-layout-responsive) {
    display: block;
    position: relative
}

.i-amphtml-layout-intrinsic {
    display: inline-block;
    position: relative;
    max-width: 100%
}

.i-amphtml-layout-container,
.i-amphtml-layout-fixed-height,
[layout=container],
[layout=fixed-height][height] {
    display: block;
    position: relative
}

.i-amphtml-layout-fill,
[layout=fill]:not(.i-amphtml-layout-fill) {
    display: block;
    overflow: hidden !important;
    position: absolute;
    top: 0;
    left: 0;
    bottom: 0;
    right: 0
}

.i-amphtml-layout-flex-item,
[layout=flex-item]:not(.i-amphtml-layout-flex-item) {
    display: block;
    position: relative;
    -webkit-box-flex: 1;
    -ms-flex: 1 1 auto;
    flex: 1 1 auto
}

.i-amphtml-layout-fluid {
    position: relative
}
```

**Comment by westonruter** [comment]
Nevertheless, since WordPress generally will only generate `amp-img` with an `intrinsic` layout (unless you manually switch the layout to be something else in the block editor, but here no `style` attribute would be present), it seems the most important scenario is to deal with just images with intrinsic layout.

**Comment by westonruter** [comment]
@randomascii I've opened a PR with a fix for the first case: #2209. A build is attached to the PR for testing.

**Comment by westonruter** [comment]
I think the override of `display:inline` in the `style` attribute is going to be the only practical fix for this. This has landed so I'll close this.

**Comment by randomascii** [comment]
It turns out this is still buggy. I hit exactly the same problem on a different one of my pages. The image styles are slightly different but it is once again the display:inline style that is causing problems. I don't know if it's a regression or a new variant. I'll put details in the other bug.

**Comment by westonruter** [comment]
As noted in another thread, updating the AMP plugin from v0.6.2 to v1.2+ should fix this: https://github.com/ampproject/amphtml/issues/23033#issuecomment-506149479
